File: landa/landa_stock/delivery_note/delivery_note.js
frappe.ui.form.on("Delivery Note", {
    setup: function (frm) {
        frm.set_query("shipping_contact", erpnext.queries.contact_query);
    },
    refresh: function (frm) {
        frm.set_query("item_code", "items", function () {
            const filters = {
                valid_from_year: ["<=", frm.doc.year_of_settlement],
                valid_to_year: [">=", frm.doc.year_of_settlement],
                is_sales_item: 1,
            };

            if (frm.doc.is_return) {
                // some items cannot be returned
                filters["cannot_be_returned"] = 0;
            }

            return {
                query: "erpnext.controllers.queries.item_query",
                filters: filters,
            };
        });

        frm.set_query("selling_price_list", function (doc) {
            return {
                filters: {
                    selling: 1,
                    company: doc.company,
                },
            };
        });
    },
    year_of_settlement: function (frm) {
        // don't prefill items if the table is already populated
        if (frm.doc.items.length >= 1 && frm.doc.items[0].item_code) return;
        landa.selling.prefill_items(frm);
    },
    before_save: function (frm) {
        frm.doc.items = landa.selling.remove_zero_qty_items(frm.doc.items);
    },
    billing_contact: function (frm) {
        landa.selling.set_contact_details(frm, "billing");
    },
});

File: landa/landa_stock/delivery_note/delivery_note.py
import frappe
from erpnext.stock.doctype.delivery_note.delivery_note import make_sales_invoice
from frappe import _
from landa.utils import update_doc


def before_validate(delivery_note, event):
	"""Set Tax Category to 'Umsatzsteuer'"""

	if (not delivery_note.tax_category) and frappe.db.exists(
		"Tax Category", "Umsatzsteuer"
	):
		delivery_note.tax_category = "Umsatzsteuer"


def validate(delivery_note, event):
	"""Validate that only returnable items are returned."""
	if delivery_note.is_return:
		for item in delivery_note.items:
			if frappe.db.get_value("Item", item.item_code, "cannot_be_returned"):
				frappe.throw(_("Item {} cannot be returned").format(item.item_name))
				return


def on_submit(delivery_note, event):
	"""Auto-create Sales Invoice for Delivery Note."""
	# 2021-09-24: disabled for now, but kept in case preferences change in the future

	# from erpnext.stock.doctype.delivery_note.delivery_note import make_sales_invoice
	#
	# if delivery_note.get('create_sales_invoice'):
	# 	sales_invoice = make_sales_invoice(delivery_note.get('name'))
	# 	sales_invoice.save()
	# 	sales_invoice.submit()
	pass


def autoname(doc, event):
	"""Create Company-specific Delivery Note name."""
	from landa.utils import get_new_name

	if doc.is_return:
		doc.name = get_new_name("RET", doc.company, "Delivery Note")
	else:
		doc.name = get_new_name("LIEF", doc.company, "Delivery Note")


@frappe.whitelist()
def make_landa_sales_invoice(source_name, target_doc=None, skip_item_mapping=False):
	source_doc = frappe.get_doc("Delivery Note", source_name)
	target_doc = make_sales_invoice(source_name, target_doc, skip_item_mapping)

	update_doc(source_doc, target_doc)

	return target_doc
